### ISSUE TITLE
'MueLu::Hierarchy::PerfUtils' is not a template

### ISSUE THREAD

**Issue by ibaned** [trilinos/Trilinos#1059]
Building at commit 4282cf1 from the `develop` branch, the Albany CDash threw the following error:

```
Trilinos/packages/muelu/src/Interface/../MueCentral/MueLu_Hierarchy_def.hpp:481:34: error: 'MueLu::Hierarchy::PerfUtils' is not a template
       GetOStream(Statistics2) << PerfUtils<SC,LO,GO,NO>::PrintMatrixInfo(*Amat, "A0", params);
```

http://my.cdash.org/viewBuildError.php?buildid=1132426

@trilinos/muelu

**Comment by tawiesn** [comment]
I think the code should be
`GetOStream(Statistics2) << MueLu::PerfUtils<SC,LO,GO,NO>::PrintMatrixInfo(*Amat, "A0", params);`

@jhux2 Can you fix this?

**Comment by aprokop** [comment]
No, I think the issue is that templates should not be there. @ibaned Could you remove the templates `<SC,LO,GO,NO>` and see if it compiles. I have patch available to push with this change, just want to confirm it fixes for you.

**Comment by aprokop** [comment]
Actually, it is slightly more involved. Here is my current patch:
```diff
From 443cd6e639db011400bb6356a59b48979e89c62d Mon Sep 17 00:00:00 2001
From: Andrey Prokopenko <prokopenkoav@ornl.gov>
Date: Fri, 10 Feb 2017 12:22:09 -0500
Subject: [PATCH] MueLu: minor update to fine level statistics

- Do not throw when Operator is not a Matrix, instead produce a warning
- Fix #1059.
---
 .../muelu/src/Interface/MueLu_HierarchyManager.hpp     | 18 +++++++++++-------
 packages/muelu/src/MueCentral/MueLu_Hierarchy_decl.hpp |  1 +
 packages/muelu/src/MueCentral/MueLu_Hierarchy_def.hpp  | 18 +++++++++++-------
 packages/muelu/src/Utils/MueLu_PerfUtils_decl.hpp      |  2 +-
 4 files changed, 24 insertions(+), 15 deletions(-)

diff --git a/packages/muelu/src/Interface/MueLu_HierarchyManager.hpp b/packages/muelu/src/Interface/MueLu_HierarchyManager.hpp
index c3e1905..1788597 100644
--- a/packages/muelu/src/Interface/MueLu_HierarchyManager.hpp
+++ b/packages/muelu/src/Interface/MueLu_HierarchyManager.hpp
@@ -152,13 +152,17 @@ namespace MueLu {
         H.EnableGraphDumping("dep_graph.dot", graphOutputLevel_);
 
       if (VerboseObject::IsPrint(Statistics2)) {
-        RCP<ParameterList> params = rcp(new ParameterList());
-        params->set("printLoadBalancingInfo", true);
-        params->set("printCommInfo",          true);
-        RCP<const Matrix> Amat = rcp_dynamic_cast<const Matrix>(Op);
-        TEUCHOS_TEST_FOR_EXCEPTION(Amat==Teuchos::null, Exceptions::RuntimeError,
-                                 "MueLu::Hierarchy::Setup(): cast to Xpetra::Matrix failed");
-        VerboseObject::GetOStream(Statistics2) << PerfUtils::PrintMatrixInfo(*Amat, "A0", params);
+        RCP<Matrix> Amat = rcp_dynamic_cast<Matrix>(Op);
+
+        if (!Amat.is_null()) {
+            RCP<ParameterList> params = rcp(new ParameterList());
+            params->set("printLoadBalancingInfo", true);
+            params->set("printCommInfo",          true);
+
+            VerboseObject::GetOStream(Statistics2) << PerfUtils::PrintMatrixInfo(*Amat, "A0", params);
+        } else {
+            VerboseObject::GetOStream(Warnings1) << "Fine level operator is not a matrix, statistics are not available" << std::endl;
+        }
       }
 
       H.SetPRrebalance(doPRrebalance_);
diff --git a/packages/muelu/src/MueCentral/MueLu_Hierarchy_decl.hpp b/packages/muelu/src/MueCentral/MueLu_Hierarchy_decl.hpp
index 09c454c..1749bc2 100644
--- a/packages/muelu/src/MueCentral/MueLu_Hierarchy_decl.hpp
+++ b/packages/muelu/src/MueCentral/MueLu_Hierarchy_decl.hpp
@@ -70,6 +70,7 @@
 #include "MueLu_Level_fwd.hpp"
 #include "MueLu_MasterList.hpp"
 #include "MueLu_NoFactory.hpp"
+#include "MueLu_PerfUtils_fwd.hpp"
 #include "MueLu_PFactory_fwd.hpp"
 #include "MueLu_RFactory_fwd.hpp"
 #include "MueLu_SmootherBase_fwd.hpp"
diff --git a/packages/muelu/src/MueCentral/MueLu_Hierarchy_def.hpp b/packages/muelu/src/MueCentral/MueLu_Hierarchy_def.hpp
index d003b46..d078b52 100644
--- a/packages/muelu/src/MueCentral/MueLu_Hierarchy_def.hpp
+++ b/packages/muelu/src/MueCentral/MueLu_Hierarchy_def.hpp
@@ -472,13 +472,17 @@ namespace MueLu {
     lib_ = A->getDomainMap()->lib();
 
     if (IsPrint(Statistics2)) {
-      RCP<ParameterList> params = rcp(new ParameterList());
-      params->set("printLoadBalancingInfo", true);
-      params->set("printCommInfo",          true);
-      RCP<const Matrix> Amat = rcp_dynamic_cast<const Matrix>(A);
-      TEUCHOS_TEST_FOR_EXCEPTION(Amat==Teuchos::null, Exceptions::RuntimeError,
-                               "MueLu::Hierarchy::Setup(): cast to Xpetra::Matrix failed");
-      GetOStream(Statistics2) << PerfUtils<SC,LO,GO,NO>::PrintMatrixInfo(*Amat, "A0", params);
+      RCP<Matrix> Amat = rcp_dynamic_cast<Matrix>(A);
+
+      if (!Amat.is_null()) {
+          RCP<ParameterList> params = rcp(new ParameterList());
+          params->set("printLoadBalancingInfo", true);
+          params->set("printCommInfo",          true);
+
+          GetOStream(Statistics2) << PerfUtils::PrintMatrixInfo(*Amat, "A0", params);
+      } else {
+          GetOStream(Warnings1) << "Fine level operator is not a matrix, statistics are not available" << std::endl;
+      }
     }
 
     RCP<const FactoryManagerBase> rcpmanager = rcpFromRef(manager);
diff --git a/packages/muelu/src/Utils/MueLu_PerfUtils_decl.hpp b/packages/muelu/src/Utils/MueLu_PerfUtils_decl.hpp
index f364d85..e60b5c3 100644
--- a/packages/muelu/src/Utils/MueLu_PerfUtils_decl.hpp
+++ b/packages/muelu/src/Utils/MueLu_PerfUtils_decl.hpp
@@ -54,7 +54,7 @@
 #include <Xpetra_Import_fwd.hpp>
 #include <Xpetra_Matrix_fwd.hpp>
 
-//#include "MueLu_Utilities_fwd.hpp"
+#include "MueLu_PerfUtils_fwd.hpp"
 
 namespace MueLu {
 // MPI helpers
-- 
2.9.3
```

**Comment by ibaned** [comment]
Actually... when I build `develop` myself it works fine... and PerfUtils is definitely a template that needs template args:

https://github.com/trilinos/Trilinos/blob/master/packages/muelu/src/Utils/MueLu_PerfUtils_decl.hpp#L72

There may be something wrong with the build configuration used by CDash...

**Comment by aprokop** [comment]
We have a system in place to typedef template arguments. PerfUtils does not actually need them (as most of our code) as long as there is a proper _fwd.hpp file inclusion.

**Comment by jhux2** [comment]
> Actually, it is slightly more involved. Here is my current patch:

Andrey,

  Thanks.  Obviously the code shouldn't throw.

Jonathan

**Comment by jhux2** [comment]
I will take this.

**Comment by ibaned** [comment]
By the way, this dashboard is under @mperego 's control, so I have limited ability to debug it. It may have something to do with using GCC 4.8.4, or some other difference in configuration between what I use and it uses.

**Comment by ibaned** [comment]
Okay, I'm learning a bit about MueLu today. This almost certainly has to do with `MueLu_UseShortNames.hpp`. Somehow its trying to use the short name in @mperego 's case but not quite succeeding, and in my build it doesn't even see the short name.

**Comment by jhux2** [comment]
This same commit also appears to have caused a seemingly unrelated [failure](http://testing.sandia.gov/cdash/viewTest.php?onlyfailed&buildid=2738292).  I reverted the commit and will investigate.

**Comment by ibaned** [comment]
For what its worth, I applied @aprokop 's patch above and my build still succeeds. Don't know how the dashboard would react to it, can't edit files there.

**Comment by jhux2** [comment]
@ibaned What SEMS versions of HDF5, Boost, and NetCDF, and PNetCDF are required?

**Comment by ibaned** [comment]
neither of these builds use SEMS, and I doubt the TPL versions will affect the outcome.

**Comment by mperego** [comment]
I can try  @aprokop 's patch on that machine. I'll keep you posted.

**Comment by tawiesn** [comment]
I agree with @ibaned. It has to do with the short names (and maybe the compiler settings). One has to use either
`GetOStream(Statistics2) << PerfUtils::PrintMatrixInfo(*Amat, "A0", params);`
using the short names (i.e., one has to properly include the fwd file in the decl.hpp file) or alternatively
`GetOStream(Statistics2) << MueLu::PerfUtils<SC,LO,GO,NO>::PrintMatrixInfo(*Amat, "A0", params);`
The latter should always work but is maybe less elegant (due to the template parameters...)

I don't see why @aprokop's patch would be necessary. Do we have cases where operator is not of type Xpetra::Matrix??

**Comment by aprokop** [comment]
@tawiesn The short names would not work because of a missing fwd declaration in PerfUtils. The patch fixes that.

 I don't know why the template one does not work (the OP error indicates that the compiler thinks that PerfUtils is a part of MueLu::Hierarchy for some reason).

We may have situations where Operator is not a Matrix, but I don't think it will got through either of the paths. The only way to set it up right now is through direct Hierarchy construction.

**Comment by tawiesn** [comment]
> @tawiesn The short names would not work because of a missing fwd declaration in PerfUtils. The patch fixes that.

I have seen that and you are right. We need the fwd inclusion.

I was referring to the case that Operator is not of type Matrix. As you said: this would mean we create the hierarchy by hand. Operator not being a matrix, but an "operator" sounds more like nonlinear multigrid or something like that. Was wondering whether there are currently plans to go in that direction. Anyway, the patch should not hurt.

**Comment by mperego** [comment]
I tried:
1. GetOStream(Statistics2) << PerfUtils::PrintMatrixInfo(*Amat, "A0", params);
2. GetOStream(Statistics2) << MueLu::PerfUtils<SC,LO,GO,NO>::PrintMatrixInfo(*Amat, "A0", params);
3.  @aprokop's patch 

All of these fix the problem.
Thanks.

**Comment by ibaned** [comment]
Item (1) on @mperego's list did not compile on my machine, so I recommend we apply @aprokop 's patch or something similarly thorough.

**Comment by tawiesn** [comment]
@mperego thanks for testing.
I agree, let's apply @aprokops patch. With respect to the error it basically implements variant 1

**Comment by tawiesn** [comment]
Obviously, @jhux2 reverted the problematic patch. So we don't have to do anything now.

**Comment by aprokop** [comment]
Fixed in caa82bf22b370bc4afb1434fb66db474a9ffb891
